src/diagnostics.rs

```diff
@@ -0,0 +1,204 @@
+#![allow(unused)]
+
+use std::fmt;
+use std::io::Write;
+use std::ops::Range;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+use backtrace::Backtrace;
+use codespan::{ByteOffset, FileMap, FileName, Span};
+use codespan_reporting::termcolor::{ColorChoice, StandardStream};
+use codespan_reporting::{Diagnostic, Label, LabelStyle, Severity};
+use url::Url;
+
+use crate::resolve::Context;
+
+pub struct Diagnostics<'a> {
+    file_map: Rc<FileMap<&'a str>>,
```

Comment:
tldr: While trying to justify what I did, I think the best solution in the long run is probably to have `Arc<Diagnostic<'a>>` with
```rust
struct Diagnostic<'a> {
   file_map: FileMap<&'a str>,
   out: Arc<Mutex<StandardStream>>,
}
```

The current design stems from older iterations, in which using `Rc` or even `Rc<RefCell>` wouldn't easily have possible. Given the current structure, I guess the best approach is to already go for `Arc<Diagnostics<'a>>` and `Arc<Mutex<StandardStream>>` with #71 and #73 (parallelization) in mind.

---

There are a multitude of issues at hand here, which resulted in me deciding for this solution. I thought about several solutions here, but this one seemed the least intrusive.

Let's start with the lifetime `'a`. The whole codebase is tainted with `'a`, because it's the lifetime of all input files' content. This stems from us wanting to use a single instance of a Backend for multiple input files. As such all input files, even the ones included at a later point, need the same lifetime `'a`. To allow that, the `Generator` is passed a `&'a Arena`, in which we allocate all file content `String`s. This gives us a `&'a str`, which we then use for everything.
The `codespan` crate expects its `CodeMap` to hold all file contents (as it only accepted `String` before brendanzab/codespan#55). In fact, the `emit` function only accepted a `CodeMap` and not a single `FileMap`. This doesn't fit our model and data layout. Thus I opened brendanzab/codespan#56 as well, which allows us to emit diagnostics given just the `FileMap`. This fits our use-case, as we are basically performing a recursive-descend for includes. As such there is no harm in having `FileMap<&'a str>`. In fact using `Rc<str>` instead of `&'a str` might result in problems with pulldown_cmark borrowing from the original `str`.
(Thinking about this I think instead of a Backend we could also accept a `BuildBackend` implementation, which allows us to create a new backend for each include, which can be generic over a different lifetime. That way we don't need the arena and also don't need to hold the `String` content after we're finished processing that file. If this is something worth investigation in your opinion @HeroicKatora, we should open an issue to keep it in mind. This might especially help with parallelization.)

To output a diagnostic, we need an output and the current `FileMap`. We need to be able to output diagnostics in the frontend, the generator, the backend and CodeGenUnits. We create a new frontend for each file, but we reuse the same generator for multiple files. The frontend is consumed by the events-iterator, which is passed as argument to functions (again due to the recursive-descend). This means that the generator doesn't easily have access to the `Diagnostics` of the frontend. Thus it needs its own instance of the diagnostics on the stack, which I moved into the existing `StackElement::Context`. This means that we need two different points where we need to access the `Diagnostics`, but one can't borrow from the other.
I agree with you that the clean way would probably be to wrap the `Diagnostics` itself in an `Rc` / `Arc`. I honestly don't know where the best place to create the `StandardStream` is, given during creation we must have access to the `Config`, and we should also probably synchronize access to the `StandardStream`, leaving us with `Arc<Mutex<StandardStream>>` in the concurrent scenario. That means we should probably create a single instance in the `Generator` and clone it for usage everywhere else.